plutus-ledger-api/test/Spec/Value.hs

```diff
@@ -0,0 +1,231 @@
+
+-- | Check equality of two compiled 'Value's directly in Haskell.
+haskellEqValue :: Value -> Value -> Bool
+haskellEqValue value1 value2 = toMap value1 Haskell.== toMap value2 where
+    toMap
+        = Map.filter (Haskell.not . Map.null)
+        . Haskell.fmap (Map.filter (Haskell./= 0))
+        . Map.fromListWith (Map.unionWith (Haskell.+))
+        . Haskell.map (Haskell.fmap $ Map.fromListWith (Haskell.+))
+        . valueToLists
+
+-- | Check whether all currencies and tokens within each of the currencies occur uniquely.
+allDistinct :: Value -> Bool
+allDistinct
+    = Haskell.and
+    . Map.fromListWith (\_ _ -> False)
+    . Haskell.map (Haskell.fmap $
+        Haskell.and . Map.fromListWith (\_ _ -> False) . Haskell.map (Haskell.fmap $ \_ -> True))
+    . valueToLists
+
+-- | Return all the pairs of elements of the given list.
+--
+-- > (x, y) `elem` pairs xs ==> fromJust (x `elemIndex` xs) <= fromJust (y `elemIndex` xs)
+--
+-- >>> pairs "abc"
+-- [('a','a'),('a','b'),('b','b'),('b','c'),('c','c')]
+pairs :: [a] -> [(a, a)]
+pairs []           = []
+pairs [x]          = [(x, x)]
+pairs (x : y : xs) = (x, x) : (x, y) : pairs (y : xs)
+
+-- | Convert each list of currencies to a 'Value', check whether those 'Value' are equal to each
+-- other and dump the costs of all the checks to a golden file.
+test_EqCurrencyList :: Haskell.String -> [[(CurrencySymbol, [(TokenName, Integer)])]] -> TestNested
+test_EqCurrencyList name currencyLists =
+    nestedGoldenVsDoc name ".stat" . Pretty.vsep $
+        let attachCode value = (value, liftCodeDef value)
+            valuesWithCodes = map (attachCode . listsToValue) currencyLists
+        in pairs valuesWithCodes Haskell.<&> \((value1, valueCode1), (value2, valueCode2)) ->
+            let eqResExp = value1 `haskellEqValue` value2
+                (eqResAct, PLC.CountingSt budget) = valueCode1 `eqValueCode` valueCode2
+            -- We need the 'allDistinct' checks, because duplicated
+            -- currencies/tokens-within-the-same-currency result in undefined behavior when
+            -- checking 'Value's for equality.
+            in if allDistinct value1 && allDistinct value2 && eqResAct /= eqResExp
+                then Haskell.error $ Haskell.intercalate "\n"
+                    [ "Error when checking equality of"
+                    , "  " Haskell.++ Haskell.show value1
+                    , "and"
+                    , "  " Haskell.++ Haskell.show value2
+                    , "Expected " Haskell.++ Haskell.show eqResExp
+                    , "But got " Haskell.++ Haskell.show eqResAct
+                    ]
+                else Pretty.group $ Pretty.pretty budget
+
+test_EqValue :: TestTree
+test_EqValue =
+    runTestNestedIn ["test", "Spec"] $
+        testNestedGhc "golden"
+            [ test_EqCurrencyList "Short" currencyListOptions
+            , test_EqCurrencyList "Long" currencyLongListOptions
+            ]
```

Comment:
I don't like these tests... They cover some interesting corner cases, but not nearly enough. I'll make a ticket for improving them.